### DISCUSSION TITLE
LFX Mentorship 2022 Spring Challenge - Proposal

### DISCUSSION THREAD

**Discussion by hydai** [WasmEdge/WasmEdge#1221] [Show and tell]
### Introduction

For those who want to apply #1063 and #1123, 
This is a pretest for the application.

Please fork this repo and complete the challenge in your repo. Once you are done, post your repo’s GitHub link, as well as a screenshot demonstrating the customized host functions, and the rust SDK for creating your test cases, as a comment under this discussion.

### Requirements

1. Fork this repo and try to add brand new host functions. You should define and implement your host functions. FYR, I have created a template in https://github.com/WasmEdge/WasmEdge/tree/challenge/proposal. You can find the details in the last four commits.
2. After you finish the first step, you will need to create a wasm file to call your host functions. If you have no ideas about how to make this, please fork and refer to this example: https://github.com/second-state/wasmedge_hostfunctionexample_interface.
3. Then, once you implement both 1. and 2. You can use `wasmedge foo.wasm` to run your application.
4. Post your repo's GitHub link including the forked WasmEdge and the wasm file or the rust interface, a screenshot of the execution result, and your comments under this discussion.

Good luck!

**Comment by wangyuan249** [comment]
Hi ~ , The following is my current progress. 

Currently, I can define my host function and register module in the project.
And I can also define the same function using rust and I take advantage of `cargo build` and `cargo publish` to put my self-define lib on `crates.io` . The following figure shows the execution result of calling the function in the custom library with the new cargo project and the result while using `wasmedge` to excute  my `call.wasm` file .

![image](https://user-images.githubusercontent.com/35907099/153748385-454a49e4-9735-4d3b-b585-32496f8a5727.png)

![image](https://user-images.githubusercontent.com/35907099/153748354-952a25b3-51df-4a92-a7d1-fb88a883d66c.png)

**Reply by hydai** [reply]
Hi @wangyuan249,
Please also append your repos' links. Thanks.

**Reply by wangyuan249** [reply]
https://github.com/wangyuan249/cargoTest

**Comment by wangyuan249** [comment]
But  what confused me is how to build connection between `ExampleMod` and  `#[link(wasm_import_module = "host_function_example")]`

![image](https://user-images.githubusercontent.com/35907099/153748993-4f9cf6f8-b0fa-4f9e-a1c6-bb1da264a687.png)

![image](https://user-images.githubusercontent.com/35907099/153749036-d2a4c935-4550-4ddd-8250-1a87b4dd74bd.png)

The  solution I can search on the web is  when we want our rust file to call  extern  C/C++ function, we need to use `gcc` to build our cpp and h file as  static link library `xxx.a`  or  Dynamic link library `xxx.so`. (Refers to Rust FFI programming). 
Then after we have registered our example module in the CLI tool, should we build the whole project in root directory or  call the module in  a particular `CLI` way ?
I would be appreciated if someone who is doing this challenge or our pre-mentor can give some guidance.  Thanks！

**Reply by hydai** [reply]
Because we use the `--target wasm32-wasi`, these functions will be translated to the external functions in the import section.
You may use `wasm2wat` to see the details.

**Comment by viscropst** [comment]
Hello ~, Sorry for my slowly progress,because I'm got stuck with `clang-cl unknown warning group '-Wcast-function-type'` error when I was trying to build Wsamedge with Visual Studio's clang compiler(version has 12 only) 2 days ago.
  So I have to get clang 13 by downloading one and pair with 'llvm-windows-wasmedge' or building one based on 'llvm-windows-wasmedge' I don't need to pair. I chose the second one ([my fork of llvm-windows-wasmedge](https://github.com/viscropst/llvm-windows-wasmedge)) and it's closed to success after adding binary of libxml2 (for solve the `no libxml2` error from llvm-mt) and its dependencies to artifacts.
-------------------------------
update: finally wasmedge clang and demo build success I'll upload more details later.
update: my fork of `llvm-windows-wasmedge` won't build `llvm-mt` to ignore problem about libxml2 and reduce the [artifact](https://github.com/viscropst/llvm-windows-wasmedge/actions/runs/1864771537) size

**Comment by viscropst** [comment]
# Summary
- [here](https://github.com/viscropst/WasmEdge/tree/challenge/proposal) is my fork of this repo
- OS: Windows 10 21H2, IDE: Visual Studio 2022 and VS Code 1.64.2, rustc: 1.58.1
 
# What I Did
- fix the tourial's compliation error
- create a host module that call with user's input and return back text by host's standard input
- create a [rust app](https://github.com/viscropst/wasmedge_hello_caller) that call the host function created and example host function

# Problem and Questions
- I'm not sure how to return back string or any other type in host functions (by ExternRef<T> with pointers?),because of this problem I just return the length of user inputed and format text.

# Screenshots
![图片](https://user-images.githubusercontent.com/16207250/154827707-cc19cbf4-0d32-4ef1-960b-e0f3c79e0629.png)
![图片](https://user-images.githubusercontent.com/16207250/154827776-8934d76e-023b-419c-be3f-226b26ae9c21.png)

**Comment by gusye1234** [comment]
## Development Environments
Env for wasmedge's building refers to this [tutorial](https://wasmedge.org/book/en/extend/build.html).
Env for rust app's building refers to this [tutorial](https://wasmedge.org/book/en/index.html).



## Related Repos

* Following the fashion of the posted template, the host functions' details are in the last 4 commits of my [fork](https://github.com/gusye1234/WasmEdge/commits/challenge/proposal).
* The rust interfaces with examples are listed under [this repo](https://github.com/gusye1234/wasm-nn-pretest)



## Details

The host functions are designed for data storage. For simplicity, the counter `Env` is implemented with the data structure containing only one single integer. Three host functions are designed for communication under `host/host_example_counter/counterfunc.*`:

* `click`: counter add 1.
* `get_count`: return the current counter's number.
* `forward_by`: add any integer to the counter.

With a rust example code below:

```rust
use wasmedge_counter_interface::*;

fn main() {
    let mut counts: u32;
    counts = get_count();
    println!("Start with counter={:?}", counts);
    println!("Click three times");
    click();
    click();
    click();
    counts = get_count();
    println!("Now counter={:?}", counts);
    println!("Counter forwarded by 3");
    forward_by(3);
    counts = get_count();
    println!("Now counter={:?}", counts);
}
```

the program will output:
<img width="1433" alt="report" src="https://user-images.githubusercontent.com/34829253/154994926-6c444e07-4e86-4b59-8f69-775cc769d533.png">

**Comment by tantan31418** [comment]
## Summary
Although knowing that the application for LFX mentorship has been closed, I would like to share the things I've learned while going through this challenge.
I'm grateful to have this chance participating in WASM and really learned a lot!
## Repos
[tantan31418/WasmEdge](https://github.com/tantan31418/WasmEdge)
[tantan31418/wasmedge_hostfunctionexample_interface(Rust Interface)](https://github.com/tantan31418/wasmedge_hostfunctionexample_interface/tree/challenge)
## What I've learned
### Environment setup
- Pull docker image and build WasmEdge on notebook
	- MBA 2020 with 1.1 GHz Quad-Core Intel Core i5
	- compile and build is very slow ~=10mins
- Install Ubuntu Server 20.04 VM on Proxmox VE
- Configure WasmEdge development environment 
- [ref1-build wasm from source](https://wasmedge.org/book/en/extend/build.html)
- Build WasmEdge
### Code trace
- Trace the last 4 commits in branch challenge/proposal
- Trace wasmedge-hostfunctionexample_interface https://github.com/second-state/wasmedge_hostfunctionexample_interface
- Trace wasm-learning examples https://github.com/second-state/wasm-learning

### Rust interface and running
- Configure Rust environment
- add wasi32 target
- Make wasmfile (byte) without modifying
- Add custom host function SetClassAddr in WasmEdge and rust interface
	- Similar to HostFuncExampleSetClassName 
- Make wasmfile again
- [ref2-cli tutorial](https://wasmedge.org/book/en/index.html)
#### encountered problem
- have WasmEdge release version and custom-build version at the same time, and used the wrong one to run

### Commit and push to github
#### rust interface
- create new branch "challenge" under "wasmedge_hostfunctionexample_interface"
- ignore items under /target/
- push to new branch
#### wasmedge
- commit and push

## Screenshots
![](https://i.imgur.com/234eZCh.png)

**Reply by alabulei1** [reply]
> 

@tantan31418 Thanks for your work. Much appreciated! We also will participate in the next LFX mentorship program. Stay tuned! You could follow the Twitter  @realwasmedge or join the discord https://discord.gg/5hPQ2H5fcg to get the latest notifications.

**Comment by zephoon** [comment]
What a pity I just finished this challenge this weekend because there is so much stuff on my work these days. 
Although my application wasn't accepted, I am still very appreciative to walk through this. 
If this application is still open, pls let me know and let me take part in it if what I do as below are correct.
Thanks!


### Repos:

[WasmEdge](https://github.com/zephoon/WasmEdge/tree/challenge/proposal)

[wasmedge_hostfunctionexample_interface](https://github.com/zephoon/wasmedge_hostfunctionexample_interface/tree/challenge)


### Screenshots

<img width="750" alt="wasm" src="https://user-images.githubusercontent.com/4590967/155879220-73b43faf-fda3-4858-a16e-05631a56abfc.png">

**Reply by alabulei1** [reply]
> What a pity I just finished this challenge this weekend because there is so much stuff on my work these days. Although my application wasn't accepted, I am still very appreciative to walk through this. If this application is still open, pls let me know and let me take part in it if what I do as below are correct. Thanks!
> 
> ### Repos:
> [WasmEdge](https://github.com/zephoon/WasmEdge/tree/challenge/proposal)
> 
> [wasmedge_hostfunctionexample_interface](https://github.com/zephoon/wasmedge_hostfunctionexample_interface/tree/challenge)
> 
> ### Screenshots
> <img alt="wasm" width="750" src="https://user-images.githubusercontent.com/4590967/155879220-73b43faf-fda3-4858-a16e-05631a56abfc.png">

Much appreciated your work on this challenge. Great! However, The LFX mentorship is finished now.  But we also will participate in the next LFX mentorship program. Stay tuned! @zephoon
